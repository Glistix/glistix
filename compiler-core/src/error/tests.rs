use super::*;
use insta::assert_snapshot;

#[test]
fn test_shell_program_not_found_error() {
    let cmds = vec![
        "erl", "erlc", "escript", "rebar3", "deno", "elixir", "node", "bun", "git",
    ];
    let oses = vec!["macos", "linux", "windows"];
    let distros = vec!["ubuntu", "unknown"];

    for cmd in &cmds {
        for os in &oses {
            match *os {
                "macos" | "windows" => {
                    let err = Error::ShellProgramNotFound {
                        program: cmd.to_string(),
                        os: os.to_string(),
                        distro: "unknown".to_string(),
                    }
                    .to_diagnostics();
                    assert_snapshot!(format!("{cmd}-{os}-unknown"), err[0].text);
                }
                "linux" => {
                    for distro in &distros {
                        let err = Error::ShellProgramNotFound {
                            program: cmd.to_string(),
                            os: os.to_string(),
                            distro: distro.to_string(),
                        }
                        .to_diagnostics();
                        assert_snapshot!(format!("{cmd}-{os}-{distro}"), err[0].text);
                    }
                }
                _ => (),
            }
        }
    }
}
