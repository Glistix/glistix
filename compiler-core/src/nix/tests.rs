use crate::ast::TypedModule;
use crate::config::PackageConfig;
use crate::type_::PRELUDE_MODULE_NAME;
use crate::{
    analyse::TargetSupport,
    build::{Origin, Target},
    nix::*,
    uid::UniqueIdGenerator,
    warning::{TypeWarningEmitter, WarningEmitter},
};
use camino::{Utf8Path, Utf8PathBuf};

mod assignments;
mod basic;
mod bit_arrays;
mod blocks;
mod bools;
mod case;
mod case_clause_guards;
mod consts;
mod custom_types;
mod externals;
mod functions;
mod lists;
mod modules;
mod numbers;
mod panic;
mod prelude;
mod records;
mod recursion;
mod results;
mod strings;
mod todo;
mod tuples;
mod type_alias;
mod use_;

pub static CURRENT_PACKAGE: &str = "thepackage";

#[macro_export]
macro_rules! assert_nix_with_multiple_imports {
    ($(($name:literal, $module_src:literal)),+; $src:literal) => {
        let compiled =
            $crate::nix::tests::compile_nix($src, vec![$((CURRENT_PACKAGE, $name, $module_src)),*]).expect("compilation failed");
        let mut output = String::from("----- SOURCE CODE\n");
        for (name, src) in [$(($name, $module_src)),*] {
            output.push_str(&format!("-- {name}.gleam\n{src}\n\n"));
        }
        output.push_str(&format!("-- main.gleam\n{}\n\n----- COMPILED NIX\n{compiled}", $src));
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    };
}

#[macro_export]
macro_rules! assert_nix {
    (($dep_package:expr, $dep_name:expr, $dep_src:expr), $src:expr $(,)?) => {{
        let compiled =
            $crate::nix::tests::compile_nix($src, vec![($dep_package, $dep_name, $dep_src)])
                .expect("compilation failed");
        let output = format!(
            "----- SOURCE CODE\n{}\n\n----- COMPILED NIX\n{}",
            $src, compiled
        );
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    }};

    (($dep_package:expr, $dep_name:expr, $dep_src:expr), $src:expr, $nix:expr $(,)?) => {{
        let compiled =
            $crate::nix::tests::compile_nix($src, Some(($dep_package, $dep_name, $dep_src)))
                .expect("compilation failed");
        let output = format!(
            "----- SOURCE CODE\n{}\n\n----- COMPILED NIX\n{}",
            $src, compiled
        );
        assert_eq!(($src, output), ($src, $nix.to_string()));
    }};

    ($src:expr $(,)?) => {{
        let output = $crate::nix::tests::compile_nix($src, vec![]).expect("compilation failed");
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    }};

    ($src:expr, $js:expr $(,)?) => {{
        let output = $crate::nix::tests::compile_nix($src, vec![]).expect("compilation failed");
        assert_eq!(($src, output), ($src, $js.to_string()));
    }};
}

#[macro_export]
macro_rules! assert_nix_error {
    ($src:expr $(,)?) => {{
        let error = $crate::nix::tests::expect_nix_error($src, vec![]);
        let output = format!("----- SOURCE CODE\n{}\n\n----- ERROR\n{}", $src, error);
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    }};
}

pub fn compile(src: &str, deps: Vec<(&str, &str, &str)>) -> TypedModule {
    let mut modules = im::HashMap::new();
    let ids = UniqueIdGenerator::new();
    // DUPE: preludeinsertion
    // TODO: Currently we do this here and also in the tests. It would be better
    // to have one place where we create all this required state for use in each
    // place.
    let _ = modules.insert(
        PRELUDE_MODULE_NAME.into(),
        crate::type_::build_prelude(&ids),
    );
    let mut direct_dependencies = std::collections::HashMap::from_iter(vec![]);

    deps.iter().for_each(|(dep_package, dep_name, dep_src)| {
        let mut dep_config = PackageConfig::default();
        dep_config.name = (*dep_package).into();
        let parsed = crate::parse::parse_module(
            Utf8PathBuf::from("test/path"),
            dep_src,
            &WarningEmitter::null(),
        )
        .expect("dep syntax error");
        let mut ast = parsed.module;
        ast.name = (*dep_name).into();
        let line_numbers = LineNumbers::new(dep_src);

        let dep = crate::analyse::ModuleAnalyzerConstructor::<()> {
            target: Target::Nix,
            ids: &ids,
            origin: Origin::Src,
            importable_modules: &modules,
            warnings: &TypeWarningEmitter::null(),
            direct_dependencies: &std::collections::HashMap::new(),
            target_support: TargetSupport::Enforced,
            package_config: &dep_config,
        }
        .infer_module(ast, line_numbers, "".into())
        .expect("should successfully infer");
        let _ = modules.insert((*dep_name).into(), dep.type_info);
        let _ = direct_dependencies.insert((*dep_package).into(), ());
    });

    let parsed =
        crate::parse::parse_module(Utf8PathBuf::from("test/path"), src, &WarningEmitter::null())
            .expect("syntax error");
    let mut ast = parsed.module;
    ast.name = "my/mod".into();
    let line_numbers = LineNumbers::new(src);
    let mut config = PackageConfig::default();
    config.name = "thepackage".into();

    crate::analyse::ModuleAnalyzerConstructor::<()> {
        target: Target::Nix,
        ids: &ids,
        origin: Origin::Src,
        importable_modules: &modules,
        warnings: &TypeWarningEmitter::null(),
        direct_dependencies: &direct_dependencies,
        target_support: TargetSupport::NotEnforced,
        package_config: &config,
    }
    .infer_module(ast, line_numbers, "".into())
    .expect("should successfully infer")
}

pub fn compile_nix(src: &str, deps: Vec<(&str, &str, &str)>) -> Result<String, crate::Error> {
    let ast = compile(src, deps);
    let line_numbers = LineNumbers::new(src);
    module(
        &ast,
        &line_numbers,
        Utf8Path::new(""),
        &"".into(),
        TargetSupport::Enforced,
    )
}

pub fn expect_nix_error(src: &str, deps: Vec<(&str, &str, &str)>) -> String {
    let error = compile_nix(src, deps).expect_err("should not compile");
    println!("er: {error:#?}");
    let better_error = match error {
        crate::Error::Nix {
            error: inner_error, ..
        } => crate::Error::Nix {
            src: src.into(),
            path: Utf8PathBuf::from("/src/nix/error.gleam"),
            error: inner_error,
        },
        _ => panic!("expected nix error, got {error:#?}"),
    };
    better_error.pretty_string()
}
