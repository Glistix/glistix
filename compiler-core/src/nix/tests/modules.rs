use crate::nix::tests::CURRENT_PACKAGE;
use crate::{assert_nix, assert_nix_with_multiple_imports};

#[test]
fn empty_module() {
    assert_nix!("", "{}\n")
}

#[test]
fn unqualified_fn_call() {
    assert_nix!(
        (CURRENT_PACKAGE, "rocket_ship", r#"pub fn launch() { 1 }"#),
        r#"import rocket_ship.{launch}
pub fn go() { launch() }
"#,
    );
}

#[test]
fn aliased_unqualified_fn_call() {
    assert_nix!(
        (CURRENT_PACKAGE, "rocket_ship", r#"pub fn launch() { 1 }"#),
        r#"import rocket_ship.{launch as boom_time}
pub fn go() { boom_time() }
"#,
    );
}

#[test]
fn multiple_unqualified_fn_call() {
    assert_nix!(
        (
            CURRENT_PACKAGE,
            "rocket_ship",
            r#"
pub fn a() { 1 }
pub fn b() { 2 }"#
        ),
        r#"import rocket_ship.{a,b as bb}
pub fn go() { a() + bb() }
"#,
    );
}

#[test]
fn constant() {
    assert_nix!(
        (CURRENT_PACKAGE, "rocket_ship", r#"pub const x = 1"#),
        r#"
import rocket_ship
pub fn go() { rocket_ship.x }
"#,
    );
}

#[test]
fn alias_aliased_constant() {
    assert_nix!(
        (CURRENT_PACKAGE, "rocket_ship", r#"pub const x = 1"#),
        r#"
import rocket_ship.{ x as y }
const z = y
"#,
    );
}

#[test]
fn renamed_module() {
    assert_nix!(
        (CURRENT_PACKAGE, "x", r#"pub const v = 1"#),
        r#"
import x as y
const z = y.v
"#,
    );
}

#[test]
fn nested_module_constant() {
    assert_nix!(
        (
            CURRENT_PACKAGE,
            "rocket_ship/launcher",
            r#"pub const x = 1"#
        ),
        r#"
import rocket_ship/launcher
pub fn go() { launcher.x }
"#,
    );
}

#[test]
fn alias_constant() {
    assert_nix!(
        (CURRENT_PACKAGE, "rocket_ship", r#"pub const x = 1"#),
        r#"
import rocket_ship as boop
pub fn go() { boop.x }
"#,
    );
}

#[test]
fn alias_fn_call() {
    assert_nix!(
        (CURRENT_PACKAGE, "rocket_ship", r#"pub fn go() { 1 }"#),
        r#"
import rocket_ship as boop
pub fn go() { boop.go() }
"#,
    );
}

#[test]
fn nested_fn_call() {
    assert_nix!(
        (CURRENT_PACKAGE, "one/two", r#"pub fn go() { 1 }"#),
        r#"import one/two
pub fn go() { two.go() }"#,
    );
}

#[test]
fn nested_nested_fn_call() {
    assert_nix!(
        (CURRENT_PACKAGE, "one/two/three", r#"pub fn go() { 1 }"#),
        r#"import one/two/three
pub fn go() { three.go() }"#,
    );
}

#[test]
fn different_package_import() {
    assert_nix!(
        ("other_package", "one", r#"pub fn go() { 1 }"#),
        r#"import one
pub fn go() { one.go() }
"#,
    );
}

#[test]
fn nested_same_package() {
    assert_nix!(
        (CURRENT_PACKAGE, "one/two/three", r#"pub fn go() { 1 }"#),
        r#"import one/two/three
pub fn go() { three.go() }
"#,
    );
}

#[test]
fn discarded_duplicate_import() {
    assert_nix_with_multiple_imports!(
        ("esa/rocket_ship", r#"pub fn go() { 1 }"#),
        ("nasa/rocket_ship", r#"pub fn go() { 1 }"#);
        r#"
import esa/rocket_ship
import nasa/rocket_ship as _nasa_rocket
pub fn go() { rocket_ship.go() }
"#
    );
}

#[test]
fn discarded_duplicate_import_with_unqualified() {
    assert_nix_with_multiple_imports!(
        ("esa/rocket_ship", r#"pub fn go() { 1 }"#),
        ("nasa/rocket_ship", r#"pub fn go() { 1 }"#);
        r#"
import esa/rocket_ship
import nasa/rocket_ship.{go} as _nasa_rocket
pub fn esa_go() { rocket_ship.go() }
pub fn nasa_go() { go() }
"#
    );
}

#[test]
fn import_with_keyword() {
    assert_nix!(
        (
            CURRENT_PACKAGE,
            "rocket_ship",
            r#"
pub const inherit = 1
pub const true = 2
"#
        ),
        r#"
import rocket_ship.{inherit, true as false}
pub fn main() {
  #(inherit, false)
}
"#
    );
}

#[test]
fn constant_module_access_with_keyword() {
    assert_nix!(
        (CURRENT_PACKAGE, "rocket_ship", r#"pub const inherit = 1"#),
        r#"
import rocket_ship
pub const variable = rocket_ship.inherit
"#,
    );
}
