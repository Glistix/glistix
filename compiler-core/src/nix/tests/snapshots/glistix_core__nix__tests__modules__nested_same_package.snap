---
source: compiler-core/src/nix/tests/modules.rs
expression: "import one/two/three\npub fn go() { three.go() }\n"
---
let
  three' = builtins.import ./../one/two/three.nix;
  
  go = {}: three'.go {};
in { inherit go; }
